includes/export/class-wc-admin-report-exporter.php

```diff
@@ -0,0 +1,185 @@
+		$num_batches = (int) ceil( $total_rows / $batch_size );
+		$start_time  = time() + 5;
+
+		// Create batches, like initial import.
+		$report_batch_args = array( $export_id, $report_type, $report_args );
+
+		if ( 0 < $num_batches ) {
+			WC_Admin_Reports_Sync::queue_batches( 1, $num_batches, self::REPORT_EXPORT_ACTION, $report_batch_args );
+		}
+
+		return $total_rows;
+	}
+
+	/**
+	 * Process a report export action.
+	 *
+	 * @param int    $page_number Page number for this action.
+	 * @param string $export_id Unique ID for report (timestamp expected).
+	 * @param string $report_type Report type. E.g. 'customers'.
+	 * @param array  $report_args Report parameters, passed to data query.
+	 * @return void
+	 */
+	public static function report_export_action( $page_number, $export_id, $report_type, $report_args ) {
+		$report_args['page'] = $page_number;
+
+		$exporter = new WC_Admin_Report_CSV_Exporter( $report_type, $report_args );
+		$exporter->set_filename( "wc-{$report_type}-report-export-{$export_id}" );
+		$exporter->generate_file();
+
+		self::update_export_percentage_complete( $report_type, $export_id, $exporter->get_percent_complete() );
+	}
+
+	/**
+	 * Generate a key to reference an export status.
+	 *
+	 * @param string $report_type Report type. E.g. 'customers'.
+	 * @param string $export_id Unique ID for report (timestamp expected).
+	 * @return string Status key.
+	 */
+	protected static function get_status_key( $report_type, $export_id ) {
+		return $report_type . ':' . $export_id;
+	}
+
+	/**
+	 * Update the completion percentage of a report export.
+	 *
+	 * @param string $report_type Report type. E.g. 'customers'.
+	 * @param string $export_id Unique ID for report (timestamp expected).
+	 * @param int    $percentage Completion percentage.
+	 * @return void
+	 */
+	public static function update_export_percentage_complete( $report_type, $export_id, $percentage ) {
+		$exports_status = get_option( self::EXPORT_STATUS_OPTION, array() );
+		$status_key     = self::get_status_key( $report_type, $export_id );
+
+		$exports_status[ $status_key ] = $percentage;
+
+		update_option( self::EXPORT_STATUS_OPTION, $exports_status );
+	}
+
+	/**
+	 * Get the completion percentage of a report export.
+	 *
+	 * @param string $report_type Report type. E.g. 'customers'.
+	 * @param string $export_id Unique ID for report (timestamp expected).
+	 * @return bool|int Completion percentage, or false if export not found.
+	 */
+	public static function get_export_percentage_complete( $report_type, $export_id ) {
+		$exports_status = get_option( self::EXPORT_STATUS_OPTION, array() );
+		$status_key     = self::get_status_key( $report_type, $export_id );
+
+		if ( isset( $exports_status[ $status_key ] ) ) {
+			return $exports_status[ $status_key ];
+		}
+
+		return false;
+	}
+
+	/**
+	 * Serve the export file.
+	 */
+	public static function download_export_file() {
+		// @todo - add nonce? (nonces are good for 24 hours)
```

Comment:
🤔 if we add in a nonce control, then we will also have to send that via an email link. I'm pretty sure I've seen that pattern other places on wpcom. So nonce + session auth seems reasonable.

Reply:
If we can punt for now, let's. I believe adding nonces could cause more problems if folks try to use stale email links.